### ISSUE TITLE
Support offset/length in places that take Range/Position

### ISSUE THREAD

**Issue by DanTup** [Microsoft/vscode#51330]
There are many places in Code that take a `Range`, `Position` or `Location` which require having a line/column. Some language servers (like Dart!) mostly use offset/lengths (and VS Code also provides offset/len in many places so it's probably not all that uncommon).

While we can convert between these easily in the editor if we have an open document, but in many cases we don't and it would be expensive to open one just for this (workspace symbols, find references, etc).

It'd be nice if there was a location type that supported either, which could be used in cases where there will be an open document at the time the info is required (for ex. the workspace symbol list opens the document before it tries to navigate to the symbol) but not at the time the data is provided.

**Comment by jrieken** [comment]
I'd say it's too late for that. And btw, for workspace symbols there is a function resolve the location lazy for exactly that reason. For things like reference search I'd assume that your service has those documents in hand anyways and that you can do the conversion there (at least that's what roslyn and typescript can do)

**Comment by DanTup** [comment]
> btw, for workspace symbols there is a function to resolve the location/range lazy - for exactly that reason

Aha, I'd overlooked that one :-)

> For things like reference search I'd assume that your service has those documents in hand anyways and that you can do the conversion there

The language server does, but that's "someone elses code". I have opened a request to provide end line/column for more things (in some cases we get start line, start column, length) but I figured that Dart might not be the only one like this and many people implementing extensions over existing language servers don't have the power to change it.

I kinda expected you wouldn't want to do it but I thought it worth raising. I think in an ideal world people implementing a language extension wouldn't require specific changes to a language server like this; but that might be easier said than done. Hopefully LSP will push people towards more common data structures anyway, so the benefits of this may become reduced further.

Thanks for the info on the workspace one though, I'll make a note to do that. Our current code does work, but we're giving the same start/end pos so if you ever do anything with this data (like you just started doing for Outline in the document one - we actually use the same code for workspace/doc) it'll be more accurate :-)
